src/arch/host/include/arch/atomic.h

```diff
@@ -32,6 +32,27 @@ static inline void arch_atomic_init(atomic_t *a, int32_t value)
 	arch_atomic_set(a, value);
 }
 
+#if defined __XCC__
+static inline int32_t arch_atomic_add(atomic_t *a, int32_t value)
+{
+	int32_t result, current;
+
+	current = arch_atomic_read(a);
+	result = current + value;
+	arch_atomic_set(a, result);
+	return current;
+}
+
+static inline int32_t arch_atomic_sub(atomic_t *a, int32_t value)
+{
+	int32_t result, current;
+
+	current = arch_atomic_read(a);
+	result = current - value;
+	arch_atomic_set(a, result);
+	return current;
```

Comment:
we realise, that these aren't atomic now, right? Unneeded because single-threaded? Maybe we need to mention it

Reply:
just to add to the above: AFAIU the current versions, using `__sync_fetch_and_add()` and `__sync_fetch_and_sub()` builtins https://gcc.gnu.org/onlinedocs/gcc/_005f_005fsync-Builtins.html is indeed atomic, this new one won't be any more. I'd *really* consider using loops here

Reply:
@lyakh Do the above new gcc atomic operations based functions look OK?

The xcc atomics are now addressed only with comment  "fake atomic functions". They should be good enough for static non-threaded version though.

Reply:
@singalsu yes, nice! The GCC versions look much more sophisticated now so they must be correct! ;-) As for xt-xcc - well... I'll stop complaining about it now, because it probably would take too long to work out something that anyway will never be used. Maybe I would at least make dead-sure it's clear that this is broken by design. Maybe something like
```
/* fake locking: obviously this does nothing, and provides no real locking */
#define lock() do {} while (0)
#define unlock() do {} while (0)
static inline int32_t arch_atomic_add(atomic_t *a, int32_t value)
{
int32_t tmp;
lock();
tmp = a->value;
a->value += value;
unlock();
return tmp;
}
```

Reply:
Great, I'll do such change for xcc generic version and split this to a separate PR.

Reply:
This is now separate PR #7291